minitests/iostandard/features/generate.py

```diff
@@ -0,0 +1,491 @@
+import os
+import random
+import json
+from collections import defaultdict
+
+try:
+    random.seed(int(os.getenv("SEED"), 16))
+except TypeError:
+    pass
+
+# =============================================================================
+
+
+def load_iob_sites(file_name):
+    """
+    Loads IOB site dump from the given CSV file.
+    """
+    iob_sites = defaultdict(lambda: [])
+
+    with open(file_name, "r") as fp:
+        fp.readline()
+
+        for line in fp:
+
+            fields = line.split(",")
+            if len(fields) != 10:
+                continue
+
+            iob_sites[fields[3]].append(
+                {
+                    "tile": fields[0],
+                    "name": fields[1],
+                    "type": fields[2],
+                    "bank": fields[4],
+                    "pkg_pin": fields[5],
+                    "is_bonded": bool(int(fields[6])),
+                    "is_clock": bool(int(fields[7])),
+                    "is_global_clock": bool(int(fields[8])),
+                    "is_vref": bool(int(fields[9])),
+                })
+
+    return iob_sites
+
+
+# =============================================================================
+
+IOBUF_NOT_ALLOWED = [
+    'HSTL_I',
+    'HSTL_I_18',
+    'SSTL18_I',
+]
+
+DIFF_MAP = {
+    'SSTL135': 'DIFF_SSTL135',
+}
+
+
+def gen_iosettings():
+    """
+    A generator function which yields all possible IO settings combintions.
+    """
+
+    IOSTANDARDS = (
+        'LVCMOS12',
+        'LVCMOS15',
+        'LVCMOS18',
+        'LVCMOS25',
+        'LVCMOS33',
+        'LVTTL',
+        'SSTL135',
+
+        # Those are available but not currently fuzzed.
+        #        'SSTL135_R',
+        #        'SSTL15',
+        #        'SSTL15_R',
+        #        'SSTL18_I',
+        #        'SSTL18_II',
+        #        'HSTL_I',
+        #        'HSTL_I_18',
+        #        'HSTL_II',
+        #        'HSTL_II_18',
+        #        'HSUL_12',
+        #        'MOBILE_DDR',
+    )
+
+    DRIVES = defaultdict(lambda: [None])
+    DRIVES.update(
+        {
+            "LVTTL": [4, 8, 12, 16, 24],
```

Comment:
Any way we could document the DRIVE strengths via a tcl output rather than coding it this way?

Reply:
Need to check that. Those values are taken from the official documentation: https://www.xilinx.com/support/documentation/user_guides/ug471_7Series_SelectIO.pdf